File: intel_briefing/pipeline/dedup.py
# ABOUTME: Deduplication logic for Intel Briefing pipeline.
# ABOUTME: Removes duplicate items by title (case-insensitive) and deduplicates politics/topics overlap.
import logging
from intel_briefing.models import IntelItem

logger = logging.getLogger(__name__)


def dedup_items(items: list[IntelItem]) -> list[IntelItem]:
    """Remove duplicate items within a list using case-insensitive title matching.

    Items with empty or missing titles are always kept (they cannot be
    meaningfully deduplicated by title).

    Args:
        items: List of IntelItem objects, possibly containing duplicates.

    Returns:
        List with duplicates removed, preserving first-occurrence order.
    """
    seen: set[str] = set()
    result: list[IntelItem] = []
    for item in items:
        key = item.title.strip().lower() if item.title else ""
        if not key:
            # Keep items with empty titles — cannot deduplicate them
            result.append(item)
            continue
        if key not in seen:
            seen.add(key)
            result.append(item)
    return result


def dedup_across_sections(
    sections: dict[str, list[IntelItem]],
) -> dict[str, list[IntelItem]]:
    """Deduplicate items across the politics and topics sections.

    If the same post (matched by id) appears in both politics and topics,
    keep it in politics and remove it from topics. This avoids double-counting
    posts from tracked political accounts that also match tracked keywords.

    Args:
        sections: Dict mapping section names to lists of IntelItem objects.

    Returns:
        The same dict with cross-section duplicates removed.
    """
    politics_ids: set[str] = {item.id for item in sections.get("politics", [])}

    if not politics_ids:
        return sections

    topics = sections.get("topics", [])
    if not topics:
        return sections

    original_count = len(topics)
    sections["topics"] = [item for item in topics if item.id not in politics_ids]
    removed = original_count - len(sections["topics"])
    if removed:
        logger.debug(
            "dedup_across_sections: removed %d item(s) from topics (already in politics)",
            removed,
        )
    return sections

File: frontend/src/lib/pipeline/dedup.ts
// ABOUTME: Deduplication logic for Info Aggregation pipeline.
// ABOUTME: Removes duplicate items by title (case-insensitive) and deduplicates social section overlap.
import type { IntelItem } from '../models'

/**
 * Remove duplicate items within a list using case-insensitive title matching.
 *
 * Items with empty or missing titles are always kept (they cannot be
 * meaningfully deduplicated by title).
 */
export function dedupItems(items: IntelItem[]): IntelItem[] {
  const seen = new Set<string>()
  const result: IntelItem[] = []

  for (const item of items) {
    const key = item.title?.trim().toLowerCase() ?? ''
    if (!key) {
      // Keep items with empty titles — cannot deduplicate them
      result.push(item)
      continue
    }
    if (!seen.has(key)) {
      seen.add(key)
      result.push(item)
    }
  }

  return result
}

/**
 * Deduplicate items within the social section.
 *
 * If the same URL appears from both accounts and topics/trends sub-sensors,
 * keep the accounts version (more specific source wins). This avoids
 * double-counting posts from tracked accounts that also match tracked keywords.
 */
export function dedupAcrossSections(
  sections: Record<string, IntelItem[]>,
): Record<string, IntelItem[]> {
  const social = sections['social'] ?? []
  if (social.length === 0) return sections

  // Within social: if the same URL appears from both accounts and topics/trends,
  // keep the accounts version (more specific source wins).
  const accountUrls = new Set(
    social.filter(item => item.id.includes('-accounts-')).map(item => item.url).filter(Boolean),
  )
  if (accountUrls.size === 0) return sections

  return {
    ...sections,
    social: social.filter(item =>
      item.id.includes('-accounts-') || !accountUrls.has(item.url),
    ),
  }
}
